src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/MarkDuplicatesSparkData.java

```diff
@@ -0,0 +1,94 @@
+package org.broadinstitute.hellbender.utils.read.markduplicates;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+public abstract class MarkDuplicatesSparkData {
+    protected final int partitionIndex;
+    protected final String name;
+
+    MarkDuplicatesSparkData(int partitionIndex, String name) {
+        this.name = name;
+        this.partitionIndex = partitionIndex;
+    }
+
+    public abstract Type getType();
+    public abstract int getScore();
+    protected abstract void serialize(Kryo kryo, Output output);
+    public abstract int getUnclippedStartPosition();
+    public abstract int getFirstRefIndex();
+    public abstract int key(final SAMFileHeader header);
+
+
+    public int getPartitionIndex(){
+      return partitionIndex;
+    }
+    public String getName() {
+      return name;
+    }
+
+    public enum Type {
```

Comment:
it's weird that we include both Fragments and EmptyFragment as fragments, and then we do an explicit class equality check after splitting by type to further subdivide.  We should probably add a forth type, FRAGMENT_PLACEHOLDER and move that check out of handleFragments()

Reply:
done